lib/concurrent/atomic/atomic_markable_reference.rb

```diff
@@ -0,0 +1,147 @@
+      end
+
+      # @!macro [attach] atomic_markable_reference_method_mark
+      #
+      #   Gets the current marked value
+      #
+      #   @return [Boolean] the current marked value
+      def mark
+        @Reference.get[1]
+      end
+      alias_method :marked?, :mark
+
+      # @!macro [attach] atomic_markable_reference_method_set
+      #
+      #   _Unconditionally_ sets to the given value of both the reference and
+      #   the mark.
+      #
+      #   @param [Object] new_val the new value
+      #   @param [Boolean] new_mark the new mark
+      #
+      #   @return [ImmutableArray] both the new value and the new mark
+      def set(new_val, new_mark)
+        @Reference.set ImmutableArray[new_val, new_mark]
+      end
+
+      # @!macro [attach] atomic_markable_reference_method_update
+      #
+      # Pass the current value and marked state to the given block, replacing it
+      # with the block's results. May retry if the value changes during the
+      # block's execution.
+      #
+      # @yield [Object] Calculate a new value and marked state for the atomic
+      #   reference using given (old) value and (old) marked
+      # @yieldparam [Object] old_val the starting value of the atomic reference
+      # @yieldparam [Boolean] old_mark the starting state of marked
+      #
+      # @return [ImmutableArray] the new value and new mark
+      def update
+        loop do
+          old_val, old_mark = @Reference.get
+          new_val, new_mark = yield old_val, old_mark
+
+          if compare_and_set old_val, new_val, old_mark, new_mark
+            return ImmutableArray[new_val, new_mark]
+          end
+        end
+      end
+
+      # @!macro [attach] atomic_markable_reference_method_try_update
+      #
+      # Pass the current value to the given block, replacing it
+      # with the block's result. Raise an exception if the update
+      # fails.
+      #
+      # @yield [Object] Calculate a new value and marked state for the atomic
+      #   reference using given (old) value and (old) marked
+      # @yieldparam [Object] old_val the starting value of the atomic reference
+      # @yieldparam [Boolean] old_mark the starting state of marked
+      #
+      # @return [ImmutableArray] the new value and marked state
+      #
+      # @raise [Concurrent::ConcurrentUpdateError] if the update fails
+      def try_update
+        old_val, old_mark = @Reference.get
+        new_val, new_mark = yield old_val, old_mark
+
+        unless compare_and_set old_val, new_val, old_mark, new_mark
+          fail ::Concurrent::ConcurrentUpdateError,
+               'AtomicMarkableReference: Update failed due to race condition.',
+               'Note: If you would like to guarantee an update, please use ' \
+               'the `AtomicMarkableReference#update` method.'
+        end
+
+        ImmutableArray[new_val, new_mark]
+      end
+
+      # Internal/private ImmutableArray for representing pairs
+      class ImmutableArray < Array
```

Comment:
Another option, instead of a frozen array (which is a little misleading since it provides a *ton* of methods that simply won't work and are completely unnecessary) is the `ImmutableStruct` class I just pushed to PR #294. This is a variation of the struct @pitr-ch spiked earlier. I haven't run the benchmarks on JRuby or Rbx yet, but it's pretty performant on MRI.

```ruby
AtomicMarkableReferenceState = Conurrent::ImmutableStruct.new(:value, :marked?)
 #=> AtomicMarkableReferenceState 
state = AtomicMarkableReferenceState.new(42, true)
 #=> #<concurrent/immutable_struct AtomicMarkableReferenceState {:value=>42, :marked?=>true}>
state.value
 #=> 42 
state.marked?
 #=> true 
```